src/main/java/io/reactivex/internal/operators/flowable/FlowableMergeWithMaybe.java

```diff
@@ -0,0 +1,359 @@
+        MergeWithObserver<T> parent = new MergeWithObserver<T>(observer);
+        observer.onSubscribe(parent);
+        source.subscribe(parent);
+        other.subscribe(parent.otherObserver);
+    }
+
+    static final class MergeWithObserver<T> extends AtomicInteger
+    implements FlowableSubscriber<T>, Subscription {
+
+        private static final long serialVersionUID = -4592979584110982903L;
+
+        final Subscriber<? super T> actual;
+
+        final AtomicReference<Subscription> mainSubscription;
+
+        final OtherObserver<T> otherObserver;
+
+        final AtomicThrowable error;
+
+        final AtomicLong requested;
+
+        final int prefetch;
+
+        final int limit;
+
+        volatile SimplePlainQueue<T> queue;
+
+        T singleItem;
+
+        volatile boolean cancelled;
+
+        volatile boolean mainDone;
+
+        volatile int otherState;
+
+        long emitted;
+
+        int consumed;
+
+        static final int OTHER_STATE_HAS_VALUE = 1;
+
+        static final int OTHER_STATE_CONSUMED_OR_EMPTY = 2;
+
+        MergeWithObserver(Subscriber<? super T> actual) {
+            this.actual = actual;
+            this.mainSubscription = new AtomicReference<Subscription>();
+            this.otherObserver = new OtherObserver<T>(this);
+            this.error = new AtomicThrowable();
+            this.requested = new AtomicLong();
+            this.prefetch = bufferSize();
+            this.limit = prefetch - (prefetch >> 2);
+        }
+
+        @Override
+        public void onSubscribe(Subscription d) {
+            if (SubscriptionHelper.setOnce(mainSubscription, d)) {
+                d.request(prefetch);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (compareAndSet(0, 1)) {
+                long e = emitted;
+                if (requested.get() != e) {
+                    SimplePlainQueue<T> q = queue;
+                    if (q == null || q.isEmpty()) {
+
+                        emitted = e + 1;
+                        actual.onNext(t);
+
+                        int c = consumed + 1;
+                        if (c == limit) {
+                            consumed = 0;
+                            mainSubscription.get().request(c);
+                        } else {
+                            consumed = c;
+                        }
+                    } else {
+                        q.offer(t);
+                    }
+                } else {
+                    SimplePlainQueue<T> q = getOrCreateQueue();
+                    q.offer(t);
+                }
+                if (decrementAndGet() == 0) {
+                    return;
+                }
+            } else {
+                SimplePlainQueue<T> q = getOrCreateQueue();
+                q.offer(t);
+                if (getAndIncrement() != 0) {
+                    return;
+                }
+            }
+            drainLoop();
```

Comment:
I'm surprised to see this call `drainLoop()` here rather than `drain` (and in `otherSuccess()`) . Are you sure access to `drainLoop` will be serialized?

Reply:
When the execution reaches this, the work-in-progress indicator is known to be non-zero as we are in the drain mode already.

Reply:
Ah yes, thanks